### PR TITLE
BUG: Fix dot and inner type/value exception failures

### LINKED ISSUE #6990: BUG: Clear errors before contructing error message that use `PyObject_Repr`
It turns out in Python3, when using debugging, they have been a bit more strict about stopping one from using `PyObject_Repr` when an exception is being handled, which sounds like a good practice as `PyObject_Repr` could raise an exception. As a result, one will see an interpreter crash if an exception is already present when running `PyObject_Repr`.

We use `PyObject_Repr` a lot when constructing exceptions as we may want to include `dtype`s or something else in the message. That being said, we need to do a better job of clearing the current exception when build the exception message. So, we need to clean up the code so that errors are cleared first by using `PyErr_Clear` before constructing the new error message. If the old error message is still needed for some reason, we need to store it somewhere else before clearing.

Also, we need to do a better job at checking to see if `PyObject_Repr` or similar is creating an exception. As `PyObject_Repr` can raise an error, we need to make sure that we are checking to see if it errored and forwarding it on instead of continuing to construct our error message.

To see an example of how this can be solved take a look at this PR ( https://github.com/numpy/numpy/pull/6988 ).

<br>
Thanks to @seberg and @charris for helping me get the needed pieces to figure out what was going wrong here.

<br>
Below is a compiled check list of problems I have found. The line numbers are based on a commit present when this issue was opened; so, code changes in those files may have shifted them. That being said, the links will use that same commit so it is a good place to start. Feel free to add others to this issue.

- [x] `PyArray_AssignArray` ( [numpy/core/src/multiarray/array_assign_array.c#L271-L284](https://github.com/numpy/numpy/tree/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/array_assign_array.c#L271-L284) )
- [x] `PyArray_AssignRawScalar` ( [numpy/core/src/multiarray/array_assign_scalar.c#L204-L217](https://github.com/numpy/numpy/tree/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/array_assign_scalar.c#L204-L217) )
- [x] `PyArray_FromArray` ( [numpy/core/src/multiarray/ctors.c#L1928-L1944](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/ctors.c#L1928-L1944) ) ( resolved in https://github.com/numpy/numpy/pull/6988 )
- [x] `convert_datetime_metadata_tuple_to_datetime_metadata` ( [numpy/core/src/multiarray/datetime.c#L1815-L1821](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/datetime.c#L1815-L1821) )
- [x] `arraydescr_setstate` ( [numpy/core/src/multiarray/descriptor.c#L2742-L2746](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/descriptor.c#L2742-L2746) )
- [x] `array_astype` ( [numpy/core/src/multiarray/methods.c#L842-L856](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/methods.c#L842-L856) )
- [x] `npyiter_check_casting` ( [numpy/core/src/multiarray/nditer_constr.c#L1297-L1312](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/nditer_constr.c#L1297-L1312) )
- [x] `npyiter_check_casting` ( [numpy/core/src/multiarray/nditer_constr.c#L1319-L1335](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/multiarray/nditer_constr.c#L1319-L1335) )
- [x] `PyUFunc_ValidateCasting` ( [numpy/core/src/umath/ufunc_type_resolution.c#L66-L80](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L66-L80) )
- [x] `PyUFunc_ValidateCasting` ( [numpy/core/src/umath/ufunc_type_resolution.c#L85-L99](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L85-L99) )
- [x] `PyUFunc_AdditionTypeResolver` ( [numpy/core/src/umath/ufunc_type_resolution.c#L748-L759](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L748-L759) )
- [x] `PyUFunc_SubtractionTypeResolver` ( [numpy/core/src/umath/ufunc_type_resolution.c#L934-L945](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L934-L945) )
- [x] `PyUFunc_MultiplicationTypeResolver` ( [numpy/core/src/umath/ufunc_type_resolution.c#L1078-L1089](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L1078-L1089) )
- [x] `PyUFunc_DivisionTypeResolver` ( [numpy/core/src/umath/ufunc_type_resolution.c#L1198-L1209](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L1198-L1209) )
- [x] `PyUFunc_DefaultLegacyInnerLoopSelector` ( [numpy/core/src/umath/ufunc_type_resolution.c#L1323-L1333](https://github.com/numpy/numpy/blob/3a9c90cbfd6057e9f124f3d31b82f8c0ad22cd20/numpy/core/src/umath/ufunc_type_resolution.c#L1323-L1333) )
- [x] `raise_binary_type_reso_error` [ufunc_type_resolution.c#L72-L90](https://github.com/numpy/numpy/blob/31e71d7ce8d447cb74b9fb83875361cf7dba4579/numpy/core/src/umath/ufunc_type_resolution.c#L72-L90)
Edit: Fixed the formatting as it appears GitHub's parser's behavior changed since last creating this list.

### PR DESCRIPTION
**Pull request by jakirkham**
Partial fix: https://github.com/numpy/numpy/issues/6990

**TL;DR:** Bug fix to catch a possible second exception while building an exception message. Includes tests for `dot` and `inner` where this bug was first discovered.

**History:** Initially, this PR began as tests for the current behavior of `dot` and `inner` when they have some mismatched types that cannot be coerced into some common working type. However, this resulted in discovering a special class of bugs in C where an exception could occur during the creation of the exception message, but was not being handled. This bug has been carefully described with a list of instances in this bug report (  https://github.com/numpy/numpy/issues/6990 ). As a result, this contains a fix for one instance of this bug and tests to verify that this instance is fixed.

### CODE REVIEW

**Comment by jakirkham** [comment]
Seeing failure on `USE_DEBUG=1` only here, as well.  Not sure what is going on. Maybe related to this ( http://bugs.python.org/issue20500 ). Though it is supposedly fixed.

**Comment by jakirkham** [comment]
This definitely feels like an interpreter bug. I tried running this with `python-dbg` version 2.7.11 and did not encounter this error.

**Comment by jakirkham** [comment]
For clarity, this is what I am seeing...

    python3-dbg: ../Objects/object.c:463: PyObject_Repr: Assertion `!PyErr_Occurred()' failed.
    ./tools/travis-test.sh: line 108:  5108 Aborted                 (core dumped) $PYTHON ../tools/test-installed-numpy.py

**Comment by seberg** [comment]
It sounds a bit like an error set without an error return (i.e. a C-function does PyErr_SetString, but then does not return NULL). The only thing I can think of is that it wants a PyErr_Clear() before setting the new error, though I don't remember that this was necessary.

**Comment by jakirkham** [comment]
What about returning `-1`? It seems there are some cases `-1` is returned instead.

**Comment by seberg** [comment]
Same as NULL in that case, there are two calling conventions for python function, those functions just don't return a python object in general.

**Comment by jakirkham** [comment]
So, I have looked, but nothing obviously sticks out to me. The only thing I can think of is `PyArray_DescrFromType` return `NULL` in some cases, but doesn't use `PyErr_SetString` in all cases.

**Comment by jakirkham** [comment]
The same can be said for `PyArray_FromAny`.

**Comment by jakirkham** [comment]
I added `PyErr_Clear` above `PyErr_SetString` and it made no difference.

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -2041,6 +2041,13 @@ def __numpy_ufunc__(self, ufunc, method, pos, inputs, **kwargs):
         assert_raises(TypeError, np.dot, b, c)
         assert_raises(TypeError, c.dot, b)
 
+    def test_dot_type_mismatch(self):
+        c = 1.
+        A = np.array((1,1), dtype='i,i')
+
+        assert_raises(ValueError, np.dot, c, A)
+        assert_raises(TypeError, np.dot, A, c)
```
Comment (jakirkham) [inline numpy/core/tests/test_multiarray.py]: So, I managed to get this built so I can debug with `python3-dbg` interactively and discovered this line will cause the strange exception. However, the line before does not cause the exception.

File: numpy/core/tests/test_multiarray.py [hunk 2]
```diff
@@ -4825,6 +4832,13 @@ def test_matmul_inplace():
 
 class TestInner(TestCase):
 
+    def test_inner_type_mismatch(self):
+        c = 1.
+        A = np.array((1,1), dtype='i,i')
+
+        assert_raises(TypeError, np.inner, c, A)
+        assert_raises(TypeError, np.inner, A, c)
```
Comment (jakirkham) [inline numpy/core/tests/test_multiarray.py]: Same story as above. This line will cause the exception, as well.

Comment (charris) [inline numpy/core/tests/test_multiarray.py]: Just to add that this error occurs with python 3.5 debug also.

**Comment by jakirkham** [comment]
So, the error in the second case looks something like this.

    TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('float64') according to the rule 'safe'

Based on what @charris has put forward, my guess is trying to create the representations of these `dtypes` to fill into the exception is causing the problem. Do either of you know where this is happening?

**Comment by jakirkham** [comment]
In short, things like this are the problem. Somehow the exception needs to be stored/cleared before doing this and then recreated afterwards. ( https://github.com/numpy/numpy/blob/97c35365beda55c6dead8c50df785eb857f843f0/numpy/core/src/multiarray/array_assign_scalar.c#L207 )

**Comment by jakirkham** [comment]
I have opened a new issue ( https://github.com/numpy/numpy/issues/6990 ) related to this problem.

**Comment by jakirkham** [comment]
So, I tried to add a bug fix, but I am seeing a segmentation fault. However, I don't see it the first time I test the `dot` or `inner` case of this form `assert_raises(TypeError, <func>, A, c)`, but only the second time. Does anything look obviously wrong to either of you with this bug fix?

**Comment by jakirkham** [comment]
Tried running `gdb` and it points me to this line ( https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/convert_datatype.c#L1010 ).

File: numpy/core/src/multiarray/ctors.c [hunk 3]
```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
     /* Raise an error if the casting rule isn't followed */
     if (!PyArray_CanCastArrayTo(arr, newtype, casting)) {
```
Comment (jakirkham) [inline numpy/core/src/multiarray/ctors.c]: After some more debugging with `gdb` apparently the segmentation fault is occurring at this line.

**Comment by jakirkham** [comment]
So, this is kind of interesting. If I run the code below, I don't encounter the segmentation fault, but I do see the error message change after the first call. Namely, it changes to the error message raised by `PyArray_MatrixProduct2` from the one that we have tried to fix here. The `ValueError` will continue to show up every time I run it. A similar thing happens with `inner`. However, if I try to reassign a value to `A` (not `B`) it will segment fault.

    >>> import numpy
    >>> import numpy as np
    >>> A = np.array((1,1), dtype='i,i')
    >>> B = np.array((1.,1.), dtype='f,f')
    >>> np.dot(A, B)
    TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('V8') according to the rule 'safe'
    >>> np.dot(A, B)
    ValueError: Cannot find a common data type.
    ...
    >>> A = np.array((1,1), dtype='i,i')

File: numpy/core/src/multiarray/ctors.c [hunk 4]
```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
...
         PyObject *errmsg;
+        PyArray_Descr *arr_descr = NULL;
+        PyObject *arr_descr_repr = NULL;
+        PyObject *newtype_repr = NULL;
 
+        PyErr_Clear();
         errmsg = PyUString_FromString("Cannot cast array data from ");
-        PyUString_ConcatAndDel(&errmsg,
-                PyObject_Repr((PyObject *)PyArray_DESCR(arr)));
+        arr_descr = PyArray_DESCR(arr);
```
Comment (jakirkham) [inline numpy/core/src/multiarray/ctors.c]: I think this may have been the cause of the segmentation faults. At the time, I didn't realize this returned a borrowed references. So, I was effectively deleting part of the array, but leaving the rest of the array around. As a result, we saw some weird behavior due to this.

**Comment by jakirkham** [comment]
Segmentation fault appears to be fixed locally. I think this will check out.

**Comment by jakirkham** [comment]
Alright, looks like it checks out. Please let me know if you guys see anything else here that needs changing.

File: numpy/core/src/multiarray/ctors.c [hunk 5]
```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
...
+        if (arr_descr == NULL) {
+            Py_DECREF(newtype);
+            Py_DECREF(errmsg);
+            return NULL;
+        }
+        arr_descr_repr = PyObject_Repr((PyObject *)arr_descr);
+        Py_DECREF(arr_descr);
```
Comment (jakirkham) [inline numpy/core/src/multiarray/ctors.c]: I think this may have been the cause of the segmentation faults. At the time, I didn't realize this returned a borrowed references. So, I was effectively deleting part of the array, but leaving the rest of the array around. As a result, we saw some weird behavior due to this.

**Comment by jakirkham** [comment]
Failures on AppVeyor ( https://github.com/numpy/numpy/issues/6991 ) had not previously occurred for this PR with exception of the segmentation fault issue (also happened on Travis), which has since been resolved. I believe these to be unrelated to the content of this PR.

**Comment by jakirkham** [comment]
As AppVeyor is merging with master, which is broken, it is currently failing the tests there. So, I ran my own AppVeyor build without this merge (this is rebased on a commit on `master` before the bad commit) and it looks like it checks out ( https://ci.appveyor.com/project/jakirkham/numpy/build/1.0.2 ).

**Comment by njsmith** [comment]
I'm starting to get really confused at what I'm looking at here -- the PR description says it's new tests, the discussion is about some gnarly C bug fixes... Maybe in the future it would work better to bundle up the test changes and code changes and big fixes into one combined PR so they stay together as a unit? Or maybe I just haven't been paying enough attention, that happens too sometimes :-)

**Comment by jakirkham** [comment]
It is pretty easy to get lost here. This is what happened.

* Made this PR ( https://github.com/numpy/numpy/pull/6968 ).
* Added simple tests for exception handling.
* Tests got weird interpreter crash in debug mode.
* Broke out the tests out in the current PR to see if they worked.
* Surprisingly still got an interpreter crash in debug mode.
* Spent some time with @seberg and @charris trying to understand weird interpreter crash.
* Figured out that this is due to one of many bugs fitting this class ( https://github.com/numpy/numpy/issues/6990 ).
* Added a bug fix here.

Sorry this is super confusing. It was never my intent to confuse anyone like this. The bug discovered was really surprising and quite subtle.

**Comment by jakirkham** [comment]
Updated the PR title and description to better explain what is going on here now.

@njsmith, is this making more sense yet? If not please just ask, this was surprisingly tricky to find. The fix ends up being straightforward.

**Comment by jakirkham** [comment]
> Maybe in the future it would work better to bundle up the test changes and code changes and big fixes into one combined PR so they stay together as a unit? Or maybe I just haven't been paying enough attention, that happens too sometimes :-)

They are. It's ok. :)

**Comment by jakirkham** [comment]
Thanks everyone for your help. In particular, thanks to @seberg and @charris for helping me locate and fix this bug.
